### PR TITLE
TA Grading: TAs can now download student files.

### LINKED ISSUE #371: Download single file or zip of entire submission directory or zip of all assigned students or zip of all students
EDITED: Bumping the priority up on this feature request.
- [x] When viewing the rubric page for specific student, next to the "popout" on each file (within the file browser pane), add a "download" link, that allows the grader to save the file to their local machine.  This is necessary to view docx files, and possibly other proprietary formats.  [ highest priority ]
- [x] When viewing the rubric page for a specific student, add a button to download a zip of all of the files (submission & results & checkout) for that student.  Maybe all versions too?  This could be in the file browser pane.  Or in the student info pane.  [ low-medium priority ]
- [x] Allow a TA to download a zip of all of their assigned students for that gradeable [ medium priority ]
- [x] Same as above, but all students.  (maybe only instructor can do this?)  [ low priority ]

ORIGINAL:  We should allow for TAs a way to download student's code from the server as right now they have to go through ssh (which many don't and just bug the instructor/lead TA) to get the files. Generally, this is done to check previous student submissions (potentially for grading, so should be able to look at past submission attempts on the UI as well) as well as when a student's code doesn't work at all, it's sometimes useful to download the student's code to run locally while doing some small debug work to check how close the student was to the right answer but didn't submit working code.

### PR DESCRIPTION
**Pull request by AndrewMRPI**
The download changes requested during last meeting were done. That is, instead of downloading the folders directly, we add a separate button that downloads the "submissions", "results" and "checkout" folders. Also implemented everything in #371.   
Closes #371 

EDITED: some edits related to issue #1172, but not closing now, needs further review, especially for peer grading

### CODE REVIEW

**Comment by MasterOdin** [comment]
So I haven't looked into this fully, but I think this actually just adds more security holes when downloading files. Looking at the `downloadFile` function, if I'm a grader in a course, I can download any file that `hwphp` has access to so long as I don't use `..` or `.` in the path. This also does not fix the security hole in `display_file` either that was talked about #1172.

While we want to be able to specify arbitrary file paths to files (as it's easy), we have to at least make sure we're anchoring them into specific directories and not just let them specify a file anywhere on the file system.

**Comment by MasterOdin** [comment]
Also, looking at this PR and #1179, you should get into the habit of starting from master when you make a branch, not some previous branch you have worked on so that you're not getting a ton of random commits from previous work, but rather just the ones that are relevant to this PR.

File: site/app/controllers/MiscController.php [hunk 1]
```diff
@@ -58,4 +72,197 @@ private function display_file() {
...
+        //Additional security
+        if (!($this->core->getUser()->accessGrading())) {
+            return false;
+        }
+        $zip_file_name = $_REQUEST['gradeable_id'] . "_" . $_REQUEST['user_id'] . "_" . date("m-d-Y") . ".zip";
+        $this->core->getOutput()->useHeader(false);
+        $this->core->getOutput()->useFooter(false);
+        $temp_dir = "/tmp";
+        //makes a random zip file name on the server
+        $temp_name = md5(uniqid($this->core->getUser()->getId(), true));
```
Comment (chameco) [inline site/app/controllers/MiscController.php]: If there's not a particular reason to take the MD5 hash here, this seems unnecessary, especially given that you're already setting passing `more_entropy = true` to `uniqid()`.

File: site/app/controllers/MiscController.php [hunk 2]
```diff
@@ -58,4 +72,197 @@ private function display_file() {
...
+        $gradeable_path = $this->core->getConfig()->getCoursePath();
+        $active_version = $gradeable->getActiveVersion();
+        $folder_names = array();
+        $folder_names[] = "submissions";
+        $folder_names[] = "results";
+        $folder_names[] = "checkout";
+        $submissions_path = FileUtils::joinPaths($gradeable_path, $folder_names[0], $gradeable->getId(), $gradeable->getUser()->getId(), $active_version);
+        $results_path = FileUtils::joinPaths($gradeable_path, $folder_names[1], $gradeable->getId(), $gradeable->getUser()->getId(), $active_version);
+        $checkout_path = FileUtils::joinPaths($gradeable_path, $folder_names[2], $gradeable->getId(), $gradeable->getUser()->getId(), $active_version);
+        chdir ($temp_dir); //changes the directory to tmps
```
Comment (chameco) [inline site/app/controllers/MiscController.php]: `chdir()` changes the working directory of the entire process - after this call, anything else that depends on the program being in its starting working directory will wail. Its probably just a good idea in general to make sure you `chdir()` back to the starting working directory afterward, perhaps obtained by using `getcwd()` before this line.

File: site/app/controllers/MiscController.php [hunk 3]
```diff
@@ -58,4 +72,197 @@ private function display_file() {
...
+        }
+
+        $temp_dir = "/tmp";
+        //makes a random zip file name on the server
+        $temp_name = md5(uniqid($this->core->getUser()->getId(), true));
+        $zip_name = $temp_dir . "/" . $temp_name . ".zip";
+        $gradeable = $this->core->getQueries()->getGradeable($_REQUEST['gradeable_id']);
+        $gradeable_path = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "submissions",
+            $gradeable->getId());
+        chdir ($temp_dir); //changes the directory to tmp
```
Comment (chameco) [inline site/app/controllers/MiscController.php]: Same comments as above w/r/t `chdir()`.

File: site/app/controllers/MiscController.php [hunk 4]
```diff
@@ -58,4 +72,197 @@ private function display_file() {
             }
         }
     }
+
+    private function downloadFile() {
+        //Additional security
+        if (!($this->core->getUser()->accessGrading())) {
```
Comment (chameco) [inline site/app/controllers/MiscController.php]: As Matt mentioned in the comments, this is not yet sufficient for security (although it is a step in the right direction). Although students are (presumably) no longer capable of downloading files, an instructor can still download arbitrary system files, including system configuration and assignment configuration/submissions from other courses. The best way to fix this is likely to treat the provided path as subdirectory of some safe directory - in conjunction with the restriction on `.` and `..` in paths, this should prevent the most blatant abuses.

File: site/app/controllers/MiscController.php (hunk 2 above)
Comment (chameco) [inline site/app/controllers/MiscController.php]: Or, looking at `ZipArchive`, there's nothing stopping you from just giving an absolute path to `open()`. This avoids the use of `chdir()` altogether, eliminating a messy and potentially buggy instance of system-global state.

File: site/app/controllers/MiscController.php (hunk 1 above)
Comment (chameco) [inline site/app/controllers/MiscController.php]: See: http://php.net/manual/en/function.uniqid.php#91126

**Pull request merged**